File: 백준/Silver/24445. 알고리즘 수업 － 너비 우선 탐색 2/알고리즘 수업 － 너비 우선 탐색 2.py
import sys
from collections import deque

n, m, start = map(int, sys.stdin.readline().split(" "))
adjList = [[] for i in range(n + 1)]
visited = [False] * (n + 1)
for line in sys.stdin.read().rstrip().split("\n"):
    a, b = map(int, line.split(" "))
    adjList[a].append(b)
    adjList[b].append(a)
for item in adjList:
    item.sort(reverse=True)
q = deque()
visited[start] = True
q.append(start)
ans = ["0"] * n
cnt = 1
while q:
    now = q.popleft()
    ans[now - 1] = str(cnt)
    cnt += 1
    for e in adjList[now]:
        if not visited[e]:
            visited[e] = True
            q.append(e)
    
sys.stdout.write("\n".join(ans))

File: 백준/Silver/24445. 알고리즘 수업 － 너비 우선 탐색 2/알고리즘 수업 － 너비 우선 탐색 2.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int start = Integer.parseInt(st.nextToken());
        List<Integer>[] adjList = new ArrayList[n + 1];
        for (int i = 1; i <= n; i++) {
            adjList[i] = new ArrayList<>();
        }
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            adjList[a].add(b);
            adjList[b].add(a);
        }
        Queue<Integer> queue = new LinkedList<>();
        boolean[] visited = new boolean[n + 1];
        int[] ans = new int[n + 1];
        int cnt = 1;
        queue.add(start);
        visited[start] = true;
        StringBuilder sb = new StringBuilder();
        while (!queue.isEmpty()) {
            int now = queue.poll();
            ans[now] = cnt++;
            adjList[now].stream().sorted(Comparator.reverseOrder()).forEach(e -> {
                if (visited[e]) {
                    return;
                }
                visited[e] = true;
                queue.add(e);
            });
        }
        for (int i = 1; i <= n; i++) {
            sb.append(ans[i]).append("\n");
        }
        System.out.print(sb);
        br.close();
    }
}
